compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/InlineCallableReferenceToLambda.kt

```diff
@@ -49,188 +45,147 @@ internal val inlineCallableReferenceToLambdaPhase = makeIrFilePhase(
 
+        return expandInlineFunctionReferenceToLambda(expression, expression.symbol.owner)
+    }
 
-            override fun visitPropertyReference(expression: IrPropertyReference): IrExpression {
-                if (expression !in inlinableCR) {
-                    return super.visitPropertyReference(expression)
-                }
+    override fun visitPropertyReference(expression: IrPropertyReference): IrExpression {
+        if (expression !in inlinableReferences) return expression
 
-                //Use getter if field is absent...
-                val field =
-                    expression.field?.owner ?: return functionReferenceToLambda(currentScope!!, expression, expression.getter!!.owner)
+        return if (expression.field?.owner == null) {
+            // Use getter if field is absent ...
+            expandInlineFunctionReferenceToLambda(expression, expression.getter!!.owner)
+        } else {
+            // ... else use field itself
+            expandInlineFieldReferenceToLambda(expression, expression.field!!.owner)
+        }
+    }
 
-                //..else use field itself
-                val irBuilder =
-                    context.createIrBuilder(currentScope!!.scope.scopeOwnerSymbol, expression.startOffset, expression.endOffset)
-                val boundReceiver = expression.dispatchReceiver ?: expression.extensionReceiver
-                return irBuilder.irBlock(expression, IrStatementOrigin.LAMBDA) {
-                    lateinit var variableForBoundReceiver: IrVariable
-                    if (boundReceiver != null) {
-                        variableForBoundReceiver = createTmpVariable(boundReceiver, BOUND_REFERENCE_RECEIVER)
+    private fun expandInlineFunctionReferenceToLambda(expression: IrCallableReference, referencedFunction: IrFunction): IrExpression {
+        val irBuilder = context.createJvmIrBuilder(currentScope!!.scope.scopeOwnerSymbol, expression.startOffset, expression.endOffset)
+        return irBuilder.irBlock(expression, IrStatementOrigin.LAMBDA) {
+
+            val parameterTypes = (expression.type as IrSimpleType).arguments.map { (it as IrTypeProjection).type }
+            val argumentTypes = parameterTypes.dropLast(1)
+
+            val typeArgumentsMap = expression.getTypeSubstitutionMap(referencedFunction)
+
+            val boundReceiver: Pair<IrValueParameter, IrExpression>? = expression.getArgumentsWithIr().singleOrNull()
+
+            val function = buildFun {
+                setSourceRange(expression)
+                origin = JvmLoweredDeclarationOrigin.FUNCTION_REFERENCE_IMPL
+                name = Name.identifier("stub_for_inlining")
+                visibility = Visibilities.LOCAL
+                returnType = referencedFunction.returnType
+                isSuspend = false
+            }.apply {
+                if (referencedFunction is IrConstructor) {
+                    copyTypeParametersFrom(referencedFunction.parentAsClass)
```

Comment:
Sorry, but actually this line is redundant cause 'expression.getTypeSubstitutionMap(referencedFunction)' abouve already look int parent class

Reply:
Sorry again, I mislead myself and you mixing declaration site here and call site below. Actually all type parameters should be specified for correctness, but still they don't actually used because of further inlining. I can't image how it could be tested. So lets just revert type coping:
```
                if (referencedFunction is IrConstructor) {
                    copyTypeParametersFrom(referencedFunction.parentAsClass)
                }
                copyTypeParametersFrom(referencedFunction)
```

Reply:
And seems they should be used for call specification (so all parameters from parent are redundant) below

Reply:
No need to apologize! I am leaning a lot on the test suite for understanding (or the illusion thereof :) and yeah, there are no tests constraining this.

Reverting the lines.

Reply:
> And seems they should be used for call specification (so all parameters from parent are redundant) below

I don't understand what you mean here. They are copied as type parameters of the lambda, and the typeSubstitution map is used to introduce type arguements to the actual call.

Reply:
```
...
                ).run {
                    irExprBody(irCall(referencedFunction).apply {
                        this@apply.descriptor.typeParameters.forEach {
                            putTypeArgument(it.index, expression.getTypeArgument(it.index))
                        }

                        var unboundIndex = 0
...
```
You mean something like this at the call site, driven by the copied parameters instead of the substitution map?

Reply:
'typeArgumentsMap' now is unused